packages/block-editor/src/hooks/text-align.js

```diff
@@ -0,0 +1,190 @@
+
+	return validTextAlignments;
+}
+
+/**
+ * Filters registered block settings, extending attributes to include `textAlign`.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	// Allow blocks to specify their own attribute definition with default values if needed.
+	if ( 'type' in ( settings.attributes?.textAlign ?? {} ) ) {
+		return settings;
+	}
+	if ( hasBlockSupport( settings, TEXT_ALIGN_SUPPORT_KEY ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			textAlign: {
+				type: 'string',
+				// Allow for '' since it is used by the `updateTextAlignment` function
+				// in toolbar controls for special cases with defined default values.
+				enum: [ ...VALID_TEXT_ALIGNMENTS, '' ],
+			},
+		};
+	}
+
+	return settings;
+}
+
+function BlockEditTextAlignmentToolbarControlsPure( {
+	name: blockName,
+	textAlign,
+	setAttributes,
+} ) {
+	const validTextAlignments = getValidTextAlignments(
+		getBlockSupport( blockName, TEXT_ALIGN_SUPPORT_KEY )
+	);
+	const blockEditingMode = useBlockEditingMode();
+	if ( ! validTextAlignments.length || blockEditingMode !== 'default' ) {
+		return null;
+	}
+
+	const textAlignmentControls = TEXT_ALIGNMENT_OPTIONS.filter( ( control ) =>
+		validTextAlignments.includes( control.align )
+	);
+
+	const updateTextAlignment = ( nextTextAlign ) => {
+		if ( ! nextTextAlign ) {
+			const blockType = getBlockType( blockName );
+			const blockDefaultTextAlign =
+				blockType?.attributes?.textAlign?.default;
+			if ( blockDefaultTextAlign ) {
+				nextTextAlign = '';
+			}
+		}
+		setAttributes( { textAlign: nextTextAlign } );
+	};
+
+	return (
+		<BlockControls group="block">
+			<AlignmentControl
+				value={ textAlign }
+				onChange={ updateTextAlignment }
+				alignmentControls={ textAlignmentControls }
+			/>
+		</BlockControls>
+	);
+}
+
+export default {
+	edit: BlockEditTextAlignmentToolbarControlsPure,
+	useBlockProps,
+	addSaveProps: addAssignedTextAlign,
+	attributeKeys: [ 'textAlign' ],
+	hasSupport( name ) {
+		return hasBlockSupport( name, TEXT_ALIGN_SUPPORT_KEY, false );
+	},
+};
+
+function useBlockProps( { name, textAlign } ) {
+	const validTextAlignments = getValidTextAlignments(
+		getBlockSupport( name, TEXT_ALIGN_SUPPORT_KEY )
+	);
+	if ( ! validTextAlignments.length ) {
+		return null;
+	}
+	const className = classnames( {
+		[ `has-text-align-${ textAlign }` ]: textAlign,
+	} );
+	return { className };
+}
+
+/**
+ * Override props assigned to save component to inject text alignment class
+ * name if block supports it.
+ *
+ * @param {Object} props      Additional props applied to save element.
+ * @param {Object} blockType  Block type.
+ * @param {Object} attributes Block attributes.
+ *
+ * @return {Object} Filtered props applied to save element.
+ */
+export function addAssignedTextAlign( props, blockType, attributes ) {
```

Comment:
As above, this should probably also handle skipping serialization for consistency etc